tokio/src/sync/notify.rs

```diff
@@ -0,0 +1,496 @@
+
+            let me = self.get_unchecked_mut();
+            (
+                &me.notify,
+                &mut me.state,
+                Pin::new_unchecked(&mut me.waiter),
+            )
+        }
+    }
+}
+
+impl Future for RecvFuture<'_> {
+    type Output = ();
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<()> {
+        use RecvState::*;
+
+        let (notify, state, mut waiter) = self.project();
+
+        'outer: loop {
+            match *state {
+                Init => {
+                    // Optimistically try acquiring a pending notification
+                    let res = notify
+                        .state
+                        .compare_exchange(NOTIFIED, EMPTY, SeqCst, SeqCst);
+
+                    if res.is_ok() {
+                        // Acquired the notification
+                        *state = Done;
+                        continue 'outer;
+                    }
+
+                    // Acquire the lock and attempt to transition to the waiting
+                    // state.
+                    let mut waiters = notify.waiters.lock().unwrap();
+
+                    // Reload the state with the lock held
+                    let mut curr = notify.state.load(SeqCst);
+
+                    // Transition the state to WAITING.
+                    loop {
+                        match curr {
+                            EMPTY => {
+                                // Transition to WAITING
+                                let res = notify
+                                    .state
+                                    .compare_exchange(EMPTY, WAITING, SeqCst, SeqCst);
+
+                                if let Err(actual) = res {
+                                    curr = actual;
+                                    continue;
+                                }
+                            }
+                            WAITING => {}
+                            NOTIFIED => {
+                                // Try consuming the notification
+                                let res = notify
+                                    .state
+                                    .compare_exchange(NOTIFIED, EMPTY, SeqCst, SeqCst);
+
+                                match res {
+                                    Ok(_) => {
+                                        // Acquired the notification
+                                        *state = Done;
+                                        continue 'outer;
+                                    }
+                                    Err(actual) => {
+                                        curr = actual;
+                                        continue;
+                                    }
+                                }
+                            }
+                            _ => unreachable!(),
+                        }
+
+                        // The state has been transitioned to waiting
+                        break;
```

Comment:
This inverted flow is real weird. It'd be easier to read this (I think) if it was inverted so that the relevant cases used `break`, and the `continue` cases just naturally continued through the loop.